File: 2023/kotlin/day01b/day01b.kt
import java.io.File

fun usage() {
    println("usage: java -jar <jarfile> <infile>")
    System.exit(1)
}

fun process(contents: String): Int {
    var sum = 0
    val digitsMap = hashMapOf(
        "0" to 0,
        "1" to 1,
        "2" to 2,
        "3" to 3,
        "4" to 4,
        "5" to 5,
        "6" to 6,
        "7" to 7,
        "8" to 8,
        "9" to 9,
        "zero" to 0,
        "one" to 1,
        "two" to 2,
        "three" to 3,
        "four" to 4,
        "five" to 5,
        "six" to 6,
        "seven" to 7,
        "eight" to 8,
        "nine" to 9
    )
    for ( line in contents.lines() ) {
        var minIndex: Int? = null
        var maxIndex: Int? = null
        var leftDigit: Int? = null
        var rightDigit: Int? = null
        if ( line.isEmpty() ) {
            break
        }
        digitsMap.keys.forEach { digit ->
            if ( line.contains(digit) ) {
                line.indexOf(digit).let { leftIndex ->
                    if (minIndex == null || leftIndex < minIndex!!) {
                        minIndex = leftIndex
                        leftDigit = digitsMap[digit]
                    }
                }
                line.lastIndexOf(digit).let { rightIndex ->
                    if (maxIndex == null || rightIndex > maxIndex!!) {
                        maxIndex = rightIndex
                        rightDigit = digitsMap[digit]
                    }
                }
            }
        }
        sum += leftDigit!! * 10 + rightDigit!!
    }
    return sum
}

fun main(args: Array<String>) {
    if ( args.size < 1 ) {
        usage()
    }
    val filename = args[0]
    val contents = File(filename).readText()
    val result = process(contents)
    println("result = $result")
}

File: 2023/javascript/day01b/day01b.js
const fs = require('fs');

function usage() {
    console.log(`usage: ${process.argv[0]} ${process.argv[1]} <file>`);
    process.exit(1);
}

function processInput(contents) {
    let lines = contents.trim().split('\n');
    let digitsMap = {
        '0': 0,
        '1': 1,
        '2': 2,
        '3': 3,
        '4': 4,
        '5': 5,
        '6': 6,
        '7': 7,
        '8': 8,
        '9': 9,
        "zero": 0,
        "one": 1,
        "two": 2,
        "three": 3,
        "four": 4,
        "five": 5,
        "six": 6,
        "seven": 7,
        "eight": 8,
        "nine": 9,
    }
    let result = 0;
    for ( let line of lines ) {
        let minIndex = null;
        let maxIndex = null;
        let leftDigit = null;
        let rightDigit = null;
        for ( let digit of Object.keys(digitsMap) ) {
            if ( line.includes(digit) ) {
                let leftIndex = line.indexOf(digit);
                if ( minIndex == null || leftIndex < minIndex ) {
                    minIndex = leftIndex;
                    leftDigit = digitsMap[digit];
                }
                let rightIndex = line.lastIndexOf(digit);
                if ( maxIndex == null || rightIndex > maxIndex ) {
                    maxIndex = rightIndex;
                    rightDigit = digitsMap[digit];
                }
            }
        }
        result += leftDigit * 10 + rightDigit;
    }
    return result;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    let filename = process.argv[2];
    let contents = fs.readFileSync(filename, 'utf8');
    let result = processInput(contents);
    console.log(`result = ${result}`);
}

main()
